File: lib/rosterParser.js
/**
 * Transform a cleaned seed document into the parsed roster object that the
 * formatter, web app, and bot consume.
 */

import { normalizeLookupKey } from './gameData.js';

const TYPE_ORDER = ['Hero', 'Core', 'Elite', 'Support', 'Air', 'Other'];

function idToTitle(id) {
  if (!id) return '';
  if (String(id).toLowerCase() === 'raynor_s_raiders') return "Raynor's Raiders";
  if (String(id).toLowerCase() === 'kerrigan_s_swarm') return "Kerrigan's Swarm";
  return String(id)
    .replace(/_/g, ' ')
    .replace(/\b\w/g, c => c.toUpperCase());
}

function getUnitSizeProfile(unitDefinition, size) {
  return size === 'large'
    ? (unitDefinition?.large ?? {})
    : (unitDefinition?.small ?? {});
}

function buildUpgradeList(unitDefinition, size, purchasedNames) {
  const allUpgrades = Object.values(unitDefinition?.upgrades ?? {}).map((upgrade) => {
    const cost = size === 'large' ? Number(upgrade?.costL ?? 0) : Number(upgrade?.costS ?? 0);
    const active = cost <= 0 || purchasedNames.has(normalizeLookupKey(upgrade?.name));
    return {
      name: upgrade?.name ?? '',
      cost,
      activation: upgrade?.activation ?? '',
      phase: upgrade?.phase ?? '',
      description: upgrade?.description ?? upgrade?.text ?? upgrade?.rules ?? '',
      linkedTo: upgrade?.linkedTo ?? '',
      active,
    };
  });

  return {
    activeUpgrades: allUpgrades.filter(upgrade => upgrade.active),
    allUpgrades,
  };
}

function toTacticalCardDetail(cardId, faction, options) {
  const fromOptions = (options.tacticalCards ?? []).find(card => card?.id === cardId);
  const fromLibrary = fromOptions ?? options.gameData?.getTacticalCard(faction, cardId);

  return {
    id: cardId,
    name: fromLibrary?.name ?? idToTitle(cardId),
    slots: fromLibrary?.slots ?? {},
    faction: typeof fromLibrary?.faction === 'string' ? fromLibrary.faction : '',
    tags: typeof fromLibrary?.tags === 'string' ? fromLibrary.tags : '',
    frontUrl: typeof fromLibrary?.frontUrl === 'string' ? fromLibrary.frontUrl : '',
    isUnique: Boolean(fromLibrary?.isUnique),
    resource: typeof fromLibrary?.resource === 'number' ? fromLibrary.resource : null,
    gasCost: typeof fromLibrary?.gasCost === 'number'
      ? fromLibrary.gasCost
      : typeof fromLibrary?.gas === 'number'
        ? fromLibrary.gas
        : typeof fromLibrary?.costGas === 'number'
          ? fromLibrary.costGas
          : typeof fromLibrary?.cost === 'number'
            ? fromLibrary.cost
            : null,
    abilities: Array.isArray(fromLibrary?.abilities)
      ? fromLibrary.abilities
      : Array.isArray(fromLibrary?.boosts)
        ? fromLibrary.boosts
          .map((boost) => {
            const name = String(boost?.name ?? '').trim();
            const text = String(boost?.description ?? boost?.text ?? '').trim();
            if (!name && !text) return null;
            return { name, text };
          })
          .filter(Boolean)
        : [],
  };
}

/**
 * @param {object} flat  Output of fetchRoster() after seed cleaning.
 * @param {object} [options]
 * @param {object} [options.gameData]
 * @param {Array<{id:string,name:string,slots:Object}>} [options.tacticalCards]
 * @returns {object}
 */
export function parseRoster(flat, options = {}) {
  if (!options.gameData) {
    throw new Error('parseRoster() requires options.gameData for cleaned seeds');
  }

  const faction = flat?.faction ?? 'Unknown';
  const rawUnits = Array.isArray(flat?.units) ? flat.units : [];

  const units = rawUnits.map((unit) => {
    const unitDefinition = options.gameData.getUnitDefinition(faction, unit);
    if (!unitDefinition) {
      throw new Error(`No unit definition found for "${unit?.name || unit?.id || 'unknown unit'}" in faction ${faction}`);
    }

    const size = unit?.size ?? 'small';
    const sizeProfile = getUnitSizeProfile(unitDefinition, size);
    const purchasedNames = new Set((unit?.purchasedUpgrades ?? []).map(normalizeLookupKey));
    const { activeUpgrades, allUpgrades } = buildUpgradeList(unitDefinition, size, purchasedNames);
    const baseCost = Number(sizeProfile?.cost ?? 0);
    const upgradeCost = activeUpgrades.reduce((sum, upgrade) => sum + Number(upgrade.cost ?? 0), 0);

    return {
      id: unitDefinition.id ?? unit?.id ?? '',
      uid: unit?.uid ?? null,
      name: unitDefinition.name ?? unit?.name ?? unit?.id ?? 'Unknown',
      type: unitDefinition.unitType ?? 'Core',
      size,
      models: Number(unit?.count ?? sizeProfile?.models ?? 1),
      supply: Number(sizeProfile?.supply ?? 0),
      baseCost,
      totalCost: baseCost + upgradeCost,
      activeUpgrades,
      allUpgrades,
      stats: {
        hp: unitDefinition?.stats?.hp ?? '-',
        armor: unitDefinition?.stats?.armor ?? '-',
        evade: unitDefinition?.stats?.evade ?? '-',
        speed: unitDefinition?.stats?.speed ?? '-',
        shield: unitDefinition?.stats?.shield ?? null,
        size: unitDefinition?.stats?.size ?? '-',
      },
      tags: unitDefinition?.tags ?? '',
    };
  }).sort((a, b) => {
    const ai = TYPE_ORDER.indexOf(a.type);
    const bi = TYPE_ORDER.indexOf(b.type);
    return (ai === -1 ? TYPE_ORDER.length : ai) - (bi === -1 ? TYPE_ORDER.length : bi);
  });

  const tacticalCardIds = Array.isArray(flat?.tacticalCards) ? flat.tacticalCards : [];
  const tacticalCardDetails = tacticalCardIds.map(id => toTacticalCardDetail(id, faction, options));
  const supply = units.reduce((sum, unit) => sum + Number(unit.supply ?? 0), 0);
  const slotKeys = [...new Set([
    ...Object.keys(flat?.slotsUsed ?? {}),
    ...Object.keys(flat?.slotsAvailable ?? {}),
  ])];

  return {
    seed: flat?.id ?? '',
    createdAt: flat?.createdAt ?? null,
    faction,
    factionCard: idToTitle(flat?.factionCardId),
    minerals: {
      used: Number(flat?.mineralsUsed ?? 0),
      limit: Number(flat?.mineralsLimit ?? 0),
    },
    gas: {
      used: Number(flat?.gasUsed ?? 0),
      limit: Number(flat?.gasLimit ?? 0),
    },
    supply,
    resources: Number(flat?.resourceTotal ?? 0),
    slots: Object.fromEntries(
      slotKeys.map((key) => [
        key,
        {
          used: Number(flat?.slotsUsed?.[key] ?? 0),
          avail: Number(flat?.slotsAvailable?.[key] ?? 0),
        },
      ])
    ),
    units,
    tacticalCards: tacticalCardDetails.map(card => card.name),
    tacticalCardDetails,
    missionIds: (flat?.missions ?? []).map(idToTitle),
  };
}

File: android/app/src/main/java/com/scadjutant/android/domain/RosterParser.kt
package com.scadjutant.android.domain

import com.scadjutant.android.model.ParsedRoster
import com.scadjutant.android.model.ResourceUsage
import com.scadjutant.android.model.ResourceShort
import com.scadjutant.android.model.RosterUnit
import com.scadjutant.android.model.RosterUpgrade
import com.scadjutant.android.model.SlotUsage
import com.scadjutant.android.model.TacticalCardDetail
import com.scadjutant.android.model.UnitStats
import com.scadjutant.android.model.UnitTypeOrder

object RosterParser {
    fun parse(
        flat: Map<String, Any?>,
        tacticalCards: List<TacticalCardDetail> = emptyList(),
    ): ParsedRoster {
        val state = flat.mapValue("state")
        val tacticalById = tacticalCards.associateBy { it.id }

        val rawUnits = state.listValue("roster").map { rawUnit ->
            val unit = rawUnit.mapValue()
            val activeIndexes = unit.intList("activeUpgrades")
            val available = unit.listValue("availableUpgrades").map { it.mapValue() }

            val activeUpgrades = activeIndexes.mapNotNull { idx ->
                val upgrade = available.getOrNull(idx) ?: return@mapNotNull null
                val cost = if (unit.stringValue("size") == "large") {
                    upgrade.intValue("costL")
                } else {
                    upgrade.intValue("costS")
                }
                RosterUpgrade(
                    name = upgrade.stringValue("name"),
                    cost = cost,
                    activation = upgrade.stringValue("activation"),
                    phase = upgrade.stringValue("phase"),
                    description = upgrade.stringValue("description", upgrade.stringValue("text", upgrade.stringValue("rules"))),
                    linkedTo = upgrade.stringValue("linkedTo"),
                    active = true,
                )
            }

            val activeSet = activeIndexes.toSet()
            val allUpgrades = available.mapIndexed { idx, upgrade ->
                val cost = if (unit.stringValue("size") == "large") {
                    upgrade.intValue("costL")
                } else {
                    upgrade.intValue("costS")
                }
                RosterUpgrade(
                    name = upgrade.stringValue("name"),
                    cost = cost,
                    activation = upgrade.stringValue("activation"),
                    phase = upgrade.stringValue("phase"),
                    description = upgrade.stringValue("description", upgrade.stringValue("text", upgrade.stringValue("rules"))),
                    linkedTo = upgrade.stringValue("linkedTo"),
                    active = activeSet.contains(idx),
                )
            }

            val upgradeCost = activeUpgrades.sumOf { it.cost }
            RosterUnit(
                id = unit.stringValue("id"),
                name = unit.stringValue("name", unit.stringValue("id", "Unknown")),
                type = unit.stringValue("unitType", "Core"),
                size = unit.stringValue("size", "small"),
                models = unit.intValue("models", 1),
                supply = unit.intValue("supply"),
                baseCost = unit.intValue("baseCost"),
                totalCost = unit.intValue("baseCost") + upgradeCost,
                activeUpgrades = activeUpgrades.sortedWith(compareByDescending<RosterUpgrade> { it.cost }.thenBy { it.name.lowercase() }),
                allUpgrades = allUpgrades.sortedWith(compareByDescending<RosterUpgrade> { it.cost }.thenBy { it.name.lowercase() }),
                stats = UnitStats(
                    hp = unit.mapValue("stats").stringValue("hp", "-"),
                    armor = unit.mapValue("stats").stringValue("armor", "-"),
                    evade = unit.mapValue("stats").stringValue("evade", "-"),
                    speed = unit.mapValue("stats").stringValue("speed", "-"),
                    shield = unit.mapValue("stats").nullableString("shield"),
                    size = unit.mapValue("stats").stringValue("size", "-"),
                ),
                tags = unit.stringValue("tags"),
            )
        }.sortedBy { unit ->
            UnitTypeOrder.indexOf(unit.type).let { idx -> if (idx >= 0) idx else Int.MAX_VALUE }
        }

        val tacticalCardDetails = state.stringList("tacticalCardIds").map { id ->
            tacticalById[id] ?: TacticalCardDetail(
                id = id,
                name = idToTitle(id),
                slots = emptyMap(),
                faction = "",
                tags = "",
                frontUrl = "",
                isUnique = false,
                resource = null,
                gasCost = null,
                abilities = emptyList(),
            )
        }

        val slotsUsed = state.mapValue("slotsUsed")
        val slotsAvail = state.mapValue("slotsAvailable")
        val slotKeys = (slotsUsed.keys + slotsAvail.keys).distinct()

        return ParsedRoster(
            seed = flat.stringValue("id"),
            createdAt = flat.nullableString("createdAt"),
            faction = state.stringValue("faction", "Unknown"),
            factionCard = idToTitle(state.stringValue("factionCardId")),
            minerals = ResourceUsage(
                used = state.intValue("mineralsUsed"),
                limit = state.intValue("mineralsLimit", 1000),
            ),
            gas = ResourceUsage(
                used = state.intValue("gasUsed"),
                limit = state.intValue("gasLimit", 100),
            ),
            supply = state.intValue("supplyUsed"),
            resources = state.intValue("resourceTotal"),
            slots = slotKeys.associateWith { key ->
                SlotUsage(
                    used = slotsUsed.intValue(key),
                    avail = slotsAvail.intValue(key),
                )
            },
            units = rawUnits,
            tacticalCards = tacticalCardDetails.map { it.name },
            tacticalCardDetails = tacticalCardDetails.sortedBy { it.name.lowercase() },
            missionIds = state.stringList("missionIds").map(::idToTitle),
        )
    }

    fun resourceShortFor(faction: String): String = ResourceShort[faction] ?: "res"

    private fun idToTitle(id: String): String {
        val normalized = id.lowercase()
        return when (normalized) {
            "raynor_s_raiders" -> "Raynor's Raiders"
            "kerrigan_s_swarm" -> "Kerrigan's Swarm"
            else -> id.split('_').joinToString(" ") { token ->
                token.replaceFirstChar { ch -> ch.uppercase() }
            }
        }
    }
}

private fun Any?.mapValue(): Map<String, Any?> = this as? Map<String, Any?> ?: emptyMap()

private fun Map<String, Any?>.mapValue(key: String): Map<String, Any?> = this[key] as? Map<String, Any?> ?: emptyMap()

private fun Map<String, Any?>.listValue(key: String): List<Any?> = this[key] as? List<Any?> ?: emptyList()

private fun Map<String, Any?>.stringValue(key: String, default: String = ""): String = when (val value = this[key]) {
    null -> default
    is String -> value
    is Number -> value.toString()
    else -> value.toString()
}

private fun Map<String, Any?>.nullableString(key: String): String? = when (val value = this[key]) {
    null -> null
    is String -> value
    else -> value.toString()
}

private fun Map<String, Any?>.intValue(key: String, default: Int = 0): Int = when (val value = this[key]) {
    is Int -> value
    is Long -> value.toInt()
    is Double -> value.toInt()
    is Float -> value.toInt()
    is String -> value.toIntOrNull() ?: default
    else -> default
}

private fun Map<String, Any?>.intList(key: String): List<Int> = listValue(key).mapNotNull { value ->
    when (value) {
        is Int -> value
        is Long -> value.toInt()
        is Double -> value.toInt()
        is String -> value.toIntOrNull()
        else -> null
    }
}

private fun Map<String, Any?>.stringList(key: String): List<String> = listValue(key).mapNotNull { value ->
    when (value) {
        is String -> value
        else -> value?.toString()
    }
}
